Plutarch/Repr/Internal.hs

```diff
@@ -0,0 +1,107 @@
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# LANGUAGE UndecidableSuperClasses #-}
+
+module Plutarch.Repr.Internal (
+  PStruct (PStruct, unPStruct),
+  PRec (PRec, unPRec),
+  pletL,
+  grecEq,
+  gstructEq,
+  StructSameRepr,
+  UnTermRec,
+  UnTermStruct,
+) where
+
+import Data.Kind (Type)
+import Data.Proxy (Proxy (Proxy))
+import Generics.SOP (All, All2, AllZipN, Code, I, K (K), LiftedCoercible, NP (Nil, (:*)), NS (S, Z), Prod, SListI, SOP (SOP), ccompare_SOP, hcliftA2, hcollapse, para_SList)
+import Generics.SOP qualified as SOP
+
+import Plutarch.Builtin.Bool
+import Plutarch.Internal.Eq
+import Plutarch.Internal.Lift
+import Plutarch.Internal.Term (S, Term, plet)
+
+newtype PStruct (struct :: [[S -> Type]]) (s :: S) = PStruct {unPStruct :: SOP (Term s) struct}
+newtype PRec (struct :: [S -> Type]) (s :: S) = PRec {unPRec :: NP (Term s) struct}
+
+newtype PLetL s r as = PLetL {unPLetL :: NP (Term s) as -> (NP (Term s) as -> Term s r) -> Term s r}
+
+pletL' :: SListI as => NP (Term s) as -> (NP (Term s) as -> Term s r) -> Term s r
+pletL' = unPLetL $ para_SList
+  (PLetL \Nil f -> f Nil)
+  \(PLetL prev) -> PLetL \(x :* xs) f -> plet x \x' ->
+    prev xs (\xs' -> f (x' :* xs'))
+
+pletL :: All SListI as => SOP (Term s) as -> (SOP (Term s) as -> Term s r) -> Term s r
+pletL (SOP (Z x)) f = pletL' x \x' -> f (SOP $ Z x')
+pletL (SOP (S xs)) f = pletL (SOP xs) \(SOP xs') -> f (SOP $ S xs')
+
+class
+  ( SOP.Generic (a s)
+  , AllZipN @Type (Prod SOP) (LiftedCoercible I (Term s)) (Code (a s)) struct
+  , AllZipN @Type (Prod SOP) (LiftedCoercible (Term s) I) struct (Code (a s))
+  ) =>
+  StructSameRepr s a struct
+instance
+  ( SOP.Generic (a s)
+  , AllZipN @Type (Prod SOP) (LiftedCoercible I (Term s)) (Code (a s)) struct
+  , AllZipN @Type (Prod SOP) (LiftedCoercible (Term s) I) struct (Code (a s))
+  ) =>
+  StructSameRepr s a struct
+
+type family UnTermRec (struct :: [Type]) :: [S -> Type] where
+  UnTermRec '[] = '[]
+  UnTermRec (Term _ a ': rest) = a ': UnTermRec rest
+
+type family UnTermStruct' (struct :: [[Type]]) :: [[S -> Type]] where
+  UnTermStruct' '[] = '[]
+  UnTermStruct' (x ': rest) = UnTermRec x ': UnTermStruct' rest
+
+type UnTermStruct x = UnTermStruct' (Code x)
+
+--------------------------------------------------------------------------------
+
+-- NOTE/TODO: These will generate large blob code, not too efficient unfortunately.
+-- We are stuck with this for SOP and Scott however. Need some benchmark
+-- reason: https://github.com/IntersectMBO/plutus/pull/5440
+
+grecEq ::
+  forall (s :: S) (struct :: [S -> Type]).
+  All PEq struct =>
+  NP (Term s) struct ->
+  NP (Term s) struct ->
+  Term s PBool
+grecEq x y =
+  let
+    go :: [Term s PBool] -> Term s PBool
+    go [] = pconstant True
+    go [x'] = x'
+    go (x' : xs') = pif x' (go xs') (pconstant False)
```

Comment:
Small optimization here! because previously it was using `#&&` and `#&&` introduces unnecessary abs and app